src/Components/Components/src/ComponentRouteData.cs

```diff
@@ -0,0 +1,36 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Components
+{
+    /// <summary>
+    /// Describes information determined during routing that specifies
+    /// the page to be displayed.
+    /// </summary>
+    public sealed class ComponentRouteData
+    {
+        /// <summary>
+        /// Constructs an instance of <see cref="ComponentRouteData"/>.
+        /// </summary>
+        /// <param name="pageComponentType">The type of the page component matching the route.</param>
+        /// <param name="pageParameters">The parameters for the page component matching the route.</param>
+        public ComponentRouteData(Type pageComponentType, IDictionary<string, object> pageParameters)
+        {
+            PageComponentType = pageComponentType ?? throw new ArgumentNullException(nameof(pageComponentType));
+            PageParameters = pageParameters ?? throw new ArgumentNullException(nameof(pageParameters));
+        }
+
+        /// <summary>
+        /// Gets the type of the page component matching the route.
+        /// </summary>
+        public Type PageComponentType { get; }
+
+        /// <summary>
+        /// Gets the parameters for the page component matching the route.
+        /// </summary>
+        public IDictionary<string, object> PageParameters { get; }
```

Comment:
`Values` or `RouteValues`? If we're going for consistency with `RouteData`.

What are the semantics of this requiring `<string, object>` - using `<string, object>` in MVC vs `<string, string>` is a huge source of headaches. Would it work to keep the route values as `<string, string>` and do conversions later?

Reply:
Yeah you're right, it would make more sense as `RouteValues`. Much better.

> Would it work to keep the route values as <string, string> and do conversions later?

The router does supply a `<string, object>` dictionary, having already done conversions. Changing things to run the conversions later doesn't fit with the layering.

Reply:
OK, I think it's fine as-is but lets be away that this will never be consistent with ASP.NET routing broadly. server-side routing doesn't perform conversions.

Reply:
Understood. I know you know the following but I'm writing it up for the understanding of anyone else following this discussion:

The key difference is that MVC has the notion of model binding for transparently coercing incoming params, whereas we definitely don't want transparent coercion for component params as the scenario is different.

Component parameters are more akin to C# method parameters - it would be horrible if the language transparently coerced method parameters in all cases. MVC/Pages parameters are specifically for data arriving in an HTTP request, so it's not equivalent.

As such, unless we added some special way of marking a given `ParameterValue` as `Coercable`, then the layering requires routing itself to deal with type conversions at its own level.

Reply:
Renamed to `RouteValues`